### PR TITLE
add deepfloyd model

### PR DESCRIPTION
**Pull request by terrancewang**

### CODE REVIEW

File: nerfstudio/generative/deepfloyd.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        # Return magniture of gradient, not the actual loss.
+        return torch.mean(gt_grad**2) ** 0.5
+
+    @staticmethod
+    @custom_bwd
+    def backward(ctx, grad):  # pylint: disable=arguments-differ
+        del grad
+        (gt_grad,) = ctx.saved_tensors
+        batch_size = len(gt_grad)
+        return gt_grad / batch_size, None
```
Comment (tancik) [inline nerfstudio/generative/deepfloyd.py:41]: Create a new utils folder in `generative/` that has this so that It can be shared between SD and deepfloyd

File: nerfstudio/generative/deepfloyd.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+class DeepFloyd(nn.Module):
```
Comment (tancik) [inline nerfstudio/generative/deepfloyd.py:43]: Add documentation for class and methods that are currently missing docs.

File: nerfstudio/generative/deepfloyd.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+    def sds_loss(
+        self,
+        text_embeddings, 
+        image, 
+        guidance_scale, 
+        grad_scaler,
+    ) -> torch.Tensor:
+        with torch.autocast(device_type="cuda", enabled=False):
+            image = F.interpolate(image.half(), (IMG_DIM, IMG_DIM), mode="bilinear", align_corners=False)
```
Comment (tancik) [inline nerfstudio/generative/deepfloyd.py:130]: Does `IMG_DIM` always need to be 64? If not, we might want to add it as an argument.

File: nerfstudio/generative/deepfloyd.py (hunk 2 above)
Comment (tancik) [inline nerfstudio/generative/deepfloyd.py:43]: Add types to all methods

**Review by tancik — COMMENTED** [review]
Thanks for adding!

File: nerfstudio/generative/deepfloyd.py (hunk 3 above)
Comment (terrancewang) [reply nerfstudio/generative/deepfloyd.py:149]: We're only using the first model of Deepfloyd for SDS, which is trained to generate 64x64 images. Not sure what the results would look like if we used a different image size though. 

the other two models in the pipeline are the super resolution models that allow it to generate higher resolution images

**Comment by terrancewang** [comment]
unsure how to resolve the last 3 pyright errors

File: nerfstudio/generative/deepfloyd.py (hunk 3 above)
Comment (tancik) [reply nerfstudio/generative/deepfloyd.py:150]: ok, feel free to ignore my comment

File: nerfstudio/generative/gradient_utils.py
Comment (tancik) [inline nerfstudio/generative/gradient_utils.py:1]: I would make the filename just `utils.py`. I can imagine more generative util functions in the future, but probably few if any `gradient` specific utils.

**Comment by terrancewang** [comment]
i tried resolving the last 3 pyright errors but im unsure how to fix the typing for them

File: nerfstudio/generative/deepfloyd.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+        assert isinstance(negative_embeds, FloatTensor)
+        assert isinstance(prompt_embeds, FloatTensor)
+        return torch.cat([negative_embeds, prompt_embeds])
+
+    def sds_loss(
+        self,
+        text_embeddings,
+        image,
+        guidance_scale,
+        grad_scaler,
```
Comment (tancik) [inline nerfstudio/generative/deepfloyd.py:136]: Missing types

**Review by tancik — APPROVED** [review]
Lgtm
